File: app/src/main/java/com/mycustomappapply/wotttoo/adapters/MessageListAdapter.java
package com.mycustomappapply.wotttoo.adapters;

import android.content.Context;
import android.os.Build;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import com.mycustomappapply.wotttoo.R;
import com.mycustomappapply.wotttoo.models.Message;
import com.mycustomappapply.wotttoo.models.User;
import com.mycustomappapply.wotttoo.models.UserMessage;
import com.mycustomappapply.wotttoo.utils.Utils;

import java.time.LocalDateTime;
import java.util.List;

public class MessageListAdapter extends RecyclerView.Adapter {
    private static final int VIEW_TYPE_MESSAGE_SENT = 1;
    private static final int VIEW_TYPE_MESSAGE_RECEIVED = 2;
    private Context mContext;
    private List<UserMessage> mMessageList;
    public MessageListAdapter(Context context, List<UserMessage> messageList) {
        mContext = context;
        mMessageList = messageList;
    }
    @Override
    public int getItemCount() {
        return mMessageList.size();
    }
    // Determines the appropriate ViewType according to the sender of the message.
    @Override
    public int getItemViewType(int position) {
       UserMessage message = (UserMessage) mMessageList.get(position);
        if (!message.getMessage().equals("")) {

            // If the current user is the sender of the message
            return VIEW_TYPE_MESSAGE_SENT;
        } else {
            // If some other user sent the message
            return VIEW_TYPE_MESSAGE_RECEIVED;
        }
    }
    // Inflates the appropriate layout according to the ViewType.
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view;
        if (viewType == VIEW_TYPE_MESSAGE_SENT) {
            view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.firsrt_user, parent, false);
            return new SentMessageHolder(view);
        } else if (viewType == VIEW_TYPE_MESSAGE_RECEIVED) {
            view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.secound_user, parent, false);
            return new ReceivedMessageHolder(view);
        }
        return null;
    }
    // Passes the message object to a ViewHolder so that the contents can be bound to UI.
    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        UserMessage message = (UserMessage) mMessageList.get(position);
        switch (holder.getItemViewType()) {
            case VIEW_TYPE_MESSAGE_SENT:
                ((SentMessageHolder) holder).bind(message);
                break;
            case VIEW_TYPE_MESSAGE_RECEIVED:
                ((ReceivedMessageHolder) holder).bind(message);
        }
    }
    private class SentMessageHolder extends RecyclerView.ViewHolder {
        TextView messageText, timeText;
        SentMessageHolder(View itemView) {
            super(itemView);
            messageText = (TextView) itemView.findViewById(R.id.text_gchat_message_me);
            timeText = (TextView) itemView.findViewById(R.id.text_gchat_date_me);
        }
        void bind(UserMessage message) {
            messageText.setText(message.getMessage());
            // Format the stored timestamp into a readable String using method.
            timeText.setText(message.getCreatedAt());
        }
    }
    private class ReceivedMessageHolder extends RecyclerView.ViewHolder {
        TextView messageText, timeText, nameText;
        ImageView profileImage;
        ReceivedMessageHolder(View itemView) {
            super(itemView);
            messageText = (TextView) itemView.findViewById(R.id.text_gchat_message_other);
            timeText = (TextView) itemView.findViewById(R.id.text_gchat_date_other);
            nameText = (TextView) itemView.findViewById(R.id.text_gchat_user_other);
          //  profileImage = (ImageView) itemView.findViewById(R.id.image_message_profile);
        }
        void bind(UserMessage message) {
            messageText.setText(message.getNickname());
            // Format the stored timestamp into a readable String using method.
            timeText.setText(message.getCreatedAt());
            nameText.setText(message.getNickname());
            // Insert the profile image from the URL into the ImageView.
         //   Utils.displayRoundImageFromUrl(mContext, message.getSender().getProfileUrl(), profileImage);
        }
    }
}

File: app/src/main/java/com/mycustomappapply/wotttoo/adapters/MessageListAdapter.kt
package com.mycustomappapply.wotttoo.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.mycustomappapply.wotttoo.R
import com.mycustomappapply.wotttoo.models.UserMessage


class MessageListAdapter(
    private val context: Context,
    private val messageList: List<UserMessage>
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        private const val VIEW_TYPE_MESSAGE_SENT = 1
        private const val VIEW_TYPE_MESSAGE_RECEIVED = 2
    }

    override fun getItemCount(): Int = messageList.size

    override fun getItemViewType(position: Int): Int {
        val message = messageList[position]
        return if (message.message.isNotEmpty()) VIEW_TYPE_MESSAGE_SENT else VIEW_TYPE_MESSAGE_RECEIVED
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = when (viewType) {
            VIEW_TYPE_MESSAGE_SENT -> LayoutInflater.from(parent.context)
                .inflate(R.layout.firsrt_user, parent, false).let { SentMessageHolder(it) }
            VIEW_TYPE_MESSAGE_RECEIVED -> LayoutInflater.from(parent.context)
                .inflate(R.layout.secound_user, parent, false).let { ReceivedMessageHolder(it) }
            else -> throw IllegalArgumentException("Unknown view type $viewType")
        }
        return view
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val message = messageList[position]
        when (holder.getItemViewType()) {
            VIEW_TYPE_MESSAGE_SENT -> (holder as SentMessageHolder).bind(message)
            VIEW_TYPE_MESSAGE_RECEIVED -> (holder as ReceivedMessageHolder).bind(message)
        }
    }

    private class SentMessageHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val messageText: TextView = itemView.findViewById(R.id.text_gchat_message_me)
        private val timeText: TextView = itemView.findViewById(R.id.text_gchat_date_me)

        fun bind(message: UserMessage) {
            messageText.text = message.message
            timeText.text = message.createdAt
        }
    }

    private class ReceivedMessageHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val messageText: TextView = itemView.findViewById(R.id.text_gchat_message_other)
        private val timeText: TextView = itemView.findViewById(R.id.text_gchat_date_other)
        private val nameText: TextView = itemView.findViewById(R.id.text_gchat_user_other)
       // private val profileImage: ImageView = itemView.findViewById(R.id.image_message_profile)

        fun bind(message: UserMessage) {
            messageText.text = message.nickname
            timeText.text = message.createdAt
            nameText.text = message.nickname
            // Utils.displayRoundImageFromUrl(context, message.getSender().getProfileUrl(), profileImage)
        }
    }
}
